### PR TITLE
Pass refine settings into the Gene component

### PR DESCRIPTION
**Pull request by orta**
Creates the real infrastructure for passing in refine settings, and having those passed onwards to metaphysics.

### CODE REVIEW

File: Example/Emission/AppDelegate.m [hunk 1]
```diff
@@ -195,9 +195,14 @@ - (UIViewController *)viewControllerForRoute:(NSString *)route;
...
-    NSString *geneID = [[route componentsSeparatedByString:@"/"] lastObject];
-    viewController = [[ARGeneComponentViewController alloc] initWithGeneID:geneID];
+  } else if ([route hasPrefix:@"/gene/"] || [route hasPrefix:@"gene/"]) {
+    NSString *geneID = [[[[route componentsSeparatedByString:@"/"] lastObject] componentsSeparatedByString:@"?"] firstObject];
+    NSURLComponents *components = [NSURLComponents componentsWithString:route];
+    NSMutableDictionary *params = [NSMutableDictionary dictionary];
+    for ( NSURLQueryItem *item in components.queryItems) {
+      params[item.name] = item.value;
+    }
+    viewController = [[ARGeneComponentViewController alloc] initWithGeneID:geneID refineSettings:params];
```
Comment (orta) [inline Example/Emission/AppDelegate.m]: hrefs from metaphysics currently are with out a "/"

File: externals/metaphysics [hunk 2]
```diff
@@ -1 +1 @@
-Subproject commit 6ecc7b5a8bc124dd807364d102b24e52a51e5ccc
+Subproject commit 986019141a0d9b5dff36322d1f3d855f96f01c21
```
Comment (orta) [inline externals/metaphysics]: ^ those updates

File: index.ios.js [hunk 3]
```diff
@@ -1,15 +1,15 @@
...
-import './lib/app_registry';
+import './lib/relay/config'
+import './lib/app_registry'
 
 export default {
   Containers,
   Components,
   Routes,
-};
+}
```
Comment (orta) [inline index.ios.js]: removes all the `;`

File: lib/app_registry.js [hunk 4]
```diff
@@ -54,7 +54,15 @@ class Gene extends RootContainer {
   constructor(props) {
     super(props)
     this.component = Containers.Gene
-    this.route = new Routes.Gene({ geneID: props.geneID })
+    if (props.refineSettings) {
+      this.route = new Routes.Gene({
+        geneID: props.geneID,
+        medium: props.refineSettings.medium,
+        price_range: props.refineSettings.price_range
+      })
```
Comment (orta) [inline lib/app_registry.js]: hrm, maybe this is better as a merge of the two settings

File: lib/components/artwork_grids/infinite_scroll_grid.js [hunk 5]
```diff
@@ -111,6 +111,7 @@ class InfiniteScrollArtworksGrid extends React.Component {
     metaphysics(query)
       .then((results) => {
         const artworks = _.get(results, this.props.queryArtworksKeypath)
+        if (artworks === undefined) { console.error('Your queryArtworksKeypath could be wrong in the infinite_scroll_grid') }
```
Comment (orta) [inline lib/components/artwork_grids/infinite_scroll_grid.js]: caught me twice

File: lib/components/buttons/flat_white.js [hunk 6]
```diff
@@ -11,7 +11,7 @@ const AnimationDuration = 250
 const AnimatedTouchable = Animated.createAnimatedComponent(TouchableHighlight)
 const AnimatedHeadline = Animated.createAnimatedComponent(Headline)
 
-export default class InvertedButton extends React.Component {
+export default class WhiteButton extends React.Component {
```
Comment (orta) [inline lib/components/buttons/flat_white.js]: Less confusion

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">The following files do not have tests: <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/app_registry.js">lib/app_registry.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/artwork_grids/artwork.js">lib/components/artwork_grids/artwork.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/artwork_grids/infinite_scroll_grid.js">lib/components/artwork_grids/infinite_scroll_grid.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/buttons/flat_white.js">lib/components/buttons/flat_white.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/gene/artworks.js">lib/components/gene/artworks.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/home/artist_rails/artist_card.js">lib/components/home/artist_rails/artist_card.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/home/artist_rails/artist_rail.js">lib/components/home/artist_rails/artist_rail.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/home/artwork_rails/artwork_rail.js">lib/components/home/artwork_rails/artwork_rail.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/home/artwork_rails/artwork_rail_header.js">lib/components/home/artwork_rails/artwork_rail_header.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/home/hero_units.js">lib/components/home/hero_units.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/components/home/search_bar.js">lib/components/home/search_bar.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/containers/artist.js">lib/containers/artist.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/containers/gene.js">lib/containers/gene.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/containers/home.js">lib/containers/home.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/native_modules/switch_board.js">lib/native_modules/switch_board.js</a>, <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/relay/config.js">lib/relay/config.js</a> &amp; <a href="https://github.com/artsy/emission/blob/43c4f47c95d1668bdefe4f04cabdafebd4967424/lib/relay/routes.js">lib/relay/routes.js</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: lib/components/buttons/flat_white.js [hunk 7]
```diff
@@ -47,7 +47,7 @@ export default class InvertedButton extends React.Component {
     })
     const styling = {
       underlayColor: (this.props.selected ? 'black' : colors['purple-regular']),
-      style: [styles.button, { backgroundColor }],
+      style: [styles.button, { backgroundColor }, this.props.style],
```
Comment (orta) [inline lib/components/buttons/flat_white.js]: lets you customise the size etc

File: lib/components/gene/artworks.js [hunk 8]
```diff
@@ -17,43 +17,11 @@ class Artworks extends React.Component {
...
-          partner {
-            name
-          }
-          href
-        }
-      }
-    }
-    `
-  }
 }
```
Comment (orta) [inline lib/components/gene/artworks.js]: I wanted this query right next to the relay fragment in Gene.js

File: lib/containers/gene.js [hunk 9]
```diff
@@ -160,12 +164,11 @@ class Gene extends React.Component {
...
     return (<View style={{ backgroundColor: 'white'}}>
         <Separator style={{marginTop:topMargin}} />
-        <View style={{flexDirection: 'row', justifyContent: 'space-between', height: 32, marginTop:8, marginBottom:8, paddingLeft: this.commonPadding, paddingRight: this.commonPadding}} >
-          <SerifText>1,xxx, works</SerifText>
-          <WhiteButton text="REFINE" style={{ marginTop:8, height: 20, width: 90}} onPress={this.refineTapped}/>
+        <View style={{flexDirection: 'row', justifyContent: 'space-between', height: 26, marginTop:12, marginBottom:12, paddingLeft: this.commonPadding, paddingRight: this.commonPadding }} >
+          <SerifText style={{ fontStyle: 'italic', marginTop:4 }}>{ this.artworkQuerySummaryString() }</SerifText>
+          {/* <WhiteButton text="REFINE" style={{ height: 26, width: 80, fontSize: 12 }} onPress={this.refineTapped}/> */}
         </View>
         <Separator/>
```
Comment (orta) [inline lib/containers/gene.js]: will do a polish run pretty soon on the UI, once I've got it all hooked up in Eigen - should clean this mess up ^

File: lib/containers/gene.js [hunk 10]
```diff
@@ -203,16 +206,72 @@ class Gene extends React.Component {
       </ParallaxScrollView>
     )
   }
+
+  artworkQuerySummaryString = () => {
+      return this.props.gene.filtered_artworks.total + " works"
```
Comment (orta) [inline lib/containers/gene.js]: talked about moving to metaphysics, will do as a later job

File: lib/relay/config.js [hunk 11]
```diff
@@ -10,6 +10,7 @@ if (Emission.useStagingEnvironment) {
   metaphysicsURL = 'https://metaphysics-staging.artsy.net'
 } else {
   metaphysicsURL = 'https://metaphysics-production.artsy.net'
+  // metaphysicsURL = 'http://localhost:5100'
```
Comment (orta) [inline lib/relay/config.js]: should probably go

File: rn-cli.config.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+
+var config = {
+  getBlacklistRE(platform) {
+    return blacklist(platform, [
+      /coverage\/.*/
+    ])
+  }
+}
+
+module.exports = config
```
Comment (orta) [inline rn-cli.config.js]: if you run tests with coverage, react native tries to transpire the reports 👯‍♂️

**Comment by orta** [comment]
OK, we now let you know what type of refine has been added at the start

**Comment by orta** [comment]
Also fixes https://github.com/artsy/emission/issues/357 https://github.com/artsy/emission/issues/355


screen shot 2016-10-18 at 20 12 48
screen shot 2016-10-18 at 20 12 44
screen shot 2016-10-18 at 20 12 36

File: Example/Emission/AppDelegate.m (hunk 1 above)
Comment (alloy) [inline Example/Emission/AppDelegate.m]: Ugh… really? :(

File: lib/app_registry.js [hunk 13]
```diff
@@ -54,7 +55,15 @@ class Gene extends RootContainer {
   constructor(props) {
     super(props)
     this.component = Containers.Gene
-    this.route = new Routes.Gene({ geneID: props.geneID })
+
+    const medium = _.get(props, 'refineSettings.medium')
+    const price_range = _.get(props, 'refineSettings.price_range')
```
Comment (alloy) [inline lib/app_registry.js]: Why are you using `get()` with key-paths here?

File: lib/containers/gene.js [hunk 14]
```diff
@@ -203,16 +211,97 @@ class Gene extends React.Component {
       </ParallaxScrollView>
     )
   }
+
+  artworkQuerySummaryString = () => {
+    let works = this.props.gene.filtered_artworks.total.toLocaleString()
+    let medium = this.props.medium === '*' ? '' : ' ・ ' + _.startCase(this.props.medium)
+    let price_range = this.props.price_range === '*-*' ? '' : ' ・ ' + this.priceRangeToHumanReadableString(this.props.price_range)
+    return `${works} works${medium}${price_range}`
+  }
```
Comment (alloy) [inline lib/containers/gene.js]: Wonder if this code can be cleaned up by conditionally adding the components to an array and joining it with `' ・ '` at the end.

**Comment by alloy** [comment]
So nice 👏

**Pull request merged**